swarm/src/lib.rs

```diff
@@ -321,72 +323,158 @@ where
         self.network.remove_listener(id)
     }
 
-    /// Initiates a new dialing attempt to the given address.
-    pub fn dial_addr(&mut self, addr: Multiaddr) -> Result<(), DialError> {
+    /// Dial a known or unknown peer.
+    ///
+    /// See also [`DialOpts`].
+    ///
+    /// ```
+    /// # use libp2p_swarm::Swarm;
+    /// # use libp2p_swarm::dial_opts::{DialOpts, PeerCondition};
+    /// # use libp2p_core::{PeerId, Transport};
+    /// # use libp2p_core::transport::dummy::DummyTransport;
+    /// # use libp2p_swarm::DummyBehaviour;
+    /// #
+    /// let mut swarm = Swarm::new(
+    ///   DummyTransport::new().boxed(),
+    ///   DummyBehaviour::default(),
+    ///   PeerId::random(),
+    /// );
+    ///
+    /// // Dial a known peer.
+    /// swarm.dial(
+    ///   DialOpts::peer_id(PeerId::random())
+    ///      .build()
+    /// );
+    ///
+    /// // Dial an unknown peer.
+    /// swarm.dial(
+    ///   DialOpts::unknown_peer_id()
+    ///      .address("/ip6/::1/tcp/12345".parse().unwrap())
+    ///      .build()
+    /// );
+    /// ```
+    pub fn dial(&mut self, opts: DialOpts) -> Result<(), DialError> {
```

Comment:
Given how common the following is:

```rust
DialOpts::unknown_peer_id().address(listen_addr).build()
```

We could consider accepting `impl Into<DialOpts>`:

```suggestion
    pub fn dial(&mut self, opts: impl Into<DialOpts>) -> Result<(), DialError> {
```

and implement:

```rust
impl From<MultiAddress> for DialOpts { }
```

where we do exactly the above.

Reply:
Totally agree we can get rid of a boilerplate code like this. Another option is to accept an `impl DialOptsBuilder` and call: `.dial(DialOpts::.address(listen_addr))`.

Reply:
> Given how common the following is:
> 
> ```rust
> DialOpts::unknown_peer_id().address(listen_addr).build()
> ```
> 
> We could consider accepting `impl Into<DialOpts>`:
> 
> and implement:
> 
> ```rust
> impl From<MultiAddress> for DialOpts { }
> ```
> 
> where we do exactly the above.

Neat! Thanks for bringing it up @thomaseizinger. I will give this a shot.

Reply:
https://github.com/libp2p/rust-libp2p/pull/2317/commits/5e3edb3bf4c9dcbd5fb5fac8ecfd0b14b956d82b implements `From<Multiaddr>` and changes calls to `unknown_peer_id().address(xxx).build()` to the address only.



> Totally agree we can get rid of a boilerplate code like this. Another option is to accept an `impl DialOptsBuilder` and call: `.dial(DialOpts::.address(listen_addr))`.

I think `From<Multiaddr>` is the more elegant way. Given that the two solve the same issue, I would prefer the earlier over the later suggestion.